packages/headless/src/controllers/core/generated-answer/headless-searchapi-generated-answer.ts

```diff
@@ -0,0 +1,171 @@
+export interface SearchAPIGeneratedAnswerProps extends GeneratedAnswerProps {}
+
+interface SubscribeStateManager {
+  engines: Record<
+    string,
+    {
+      abortController: AbortController | undefined;
+      lastRequestId: string;
+      lastStreamId: string;
+    }
+  >;
+  getIsStreamInProgress: (genQaEngineId: string) => boolean;
+  setAbortControllerRef: (ref: AbortController, genQaEngineId: string) => void;
+  subscribeToSearchRequests: (
+    engine: CoreEngine<
+      GeneratedAnswerSection & SearchSection & DebugSection,
+      ClientThunkExtraArguments<GeneratedAnswerAPIClient>
+    >
+  ) => Unsubscribe;
+}
+
+const subscribeStateManager: SubscribeStateManager = {
+  engines: {},
+
+  setAbortControllerRef: (ref: AbortController, genQaEngineId: string) => {
+    subscribeStateManager.engines[genQaEngineId].abortController = ref;
+  },
+
+  getIsStreamInProgress: (genQaEngineId: string) => {
+    if (
+      !subscribeStateManager.engines[genQaEngineId].abortController ||
+      subscribeStateManager.engines[genQaEngineId].abortController?.signal
+        .aborted
+    ) {
+      subscribeStateManager.engines[genQaEngineId].abortController = undefined;
+      return false;
+    }
+    return true;
+  },
+
+  subscribeToSearchRequests: (engine) => {
+    const strictListener = () => {
+      const state = engine.state;
+      const requestId = state.search.requestId;
+      const streamId =
+        state.search.extendedResults.generativeQuestionAnsweringId;
+
+      const genQaEngineId = state.generatedAnswer.id;
+
+      if (
+        subscribeStateManager.engines[genQaEngineId].lastRequestId !== requestId
+      ) {
+        subscribeStateManager.engines[genQaEngineId].lastRequestId = requestId;
+        subscribeStateManager.engines[genQaEngineId].abortController?.abort();
+        engine.dispatch(resetAnswer());
+      }
+
+      const isStreamInProgress =
+        subscribeStateManager.getIsStreamInProgress(genQaEngineId);
+      if (
+        !isStreamInProgress &&
+        streamId &&
+        streamId !== subscribeStateManager.engines[genQaEngineId].lastStreamId
+      ) {
+        subscribeStateManager.engines[genQaEngineId].lastStreamId = streamId;
+        engine.dispatch(
+          streamAnswer({
+            setAbortControllerRef: (ref: AbortController) =>
+              subscribeStateManager.setAbortControllerRef(ref, genQaEngineId),
+          })
+        );
+      }
+    };
+    return engine.subscribe(strictListener);
+  },
+};
+
+export interface SearchAPIGeneratedAnswerAnalyticsClient
+  extends GeneratedAnswerAnalyticsClient {}
+
+/**
+ * Creates a `GeneratedAnswer` controller instance using the search api stream pattern.
+ *
+ * @param engine - The headless engine.
+ * @param props - The configurable `GeneratedAnswer` properties.
+ * @returns A `GeneratedAnswer` controller instance.
+ */
+export function buildSearchAPIGeneratedAnswer(
+  engine: SearchEngine,
```

Comment:
```suggestion
  engine: SearchEngine | InsightEngine,
```

and import it

Reply:
Thanks a lot for the narrowing @fbeaudoincoveo . This is indeed exactly what I needed